File: ALittle/Module/ALittleIDE/Other/GameLibrary/CEngine/Script/CEngine/UISystem/Complex/SpritePlay.lua
-- ALittle Generate Lua
module("ALittle", package.seeall)

local ___thispath = select('1', ...):match("(.+[/\\]).+$") or ""
local ___rawset = rawset
local ___pairs = pairs
local ___ipairs = ipairs


assert(ALittle.Sprite, " extends class:ALittle.Sprite is nil")
SpritePlay = Lua.Class(ALittle.Sprite, "ALittle.SpritePlay")

function SpritePlay:Ctor(ctrl_sys)
	___rawset(self, "_interval", 0)
	___rawset(self, "_play_index", 0)
end

function SpritePlay.__setter:interval(value)
	if self._interval == value then
		return
	end
	self._interval = value
	self:Play()
end

function SpritePlay.__getter:interval()
	return self._interval
end

function SpritePlay:Play()
	if self._play_loop ~= nil then
		A_LoopSystem:RemoveUpdater(self._play_loop)
		self._play_loop = nil
	end
	self._play_index = 0
	self._row_index = 1
	self._col_index = 1
	self._play_loop = LoopFunction(Lua.Bind(self.PlayUpdate, self), -1, self._interval, 0)
	A_LoopSystem:AddUpdater(self._play_loop)
end

function SpritePlay:Stop()
	if self._play_loop ~= nil then
		A_LoopSystem:RemoveUpdater(self._play_loop)
		self._play_loop = nil
	end
end

function SpritePlay:PlayUpdate()
	local total_count = self._row_count * self._col_count
	self._play_index = self._play_index + 1
	self._play_index = self._play_index % total_count
	self.row_index = Math_Floor(self._play_index / self._row_count) + 1
	self.col_index = self._play_index % self._col_count + 1
end

File: ALittleIDE/Module/ALittleIDE/Other/GameLibrary/CEngine/JSScript/CEngine/UISystem/Complex/SpritePlay.js
{
if (typeof ALittle === "undefined") window.ALittle = {};


if (ALittle.Sprite === undefined) throw new Error(" extends class:ALittle.Sprite is undefined");
ALittle.SpritePlay = JavaScript.Class(ALittle.Sprite, {
	Ctor : function(ctrl_sys) {
		this._interval = 0;
		this._play_index = 0;
	},
	set interval(value) {
		if (this._interval === value) {
			return;
		}
		this._interval = value;
		this.Play();
	},
	get interval() {
		return this._interval;
	},
	Play : function() {
		if (this._play_loop !== undefined) {
			A_WeakLoopSystem.RemoveUpdater(this._play_loop);
			this._play_loop = undefined;
		}
		this._play_index = 0;
		this._row_index = 1;
		this._col_index = 1;
		this._play_loop = ALittle.NewObject(ALittle.LoopFunction, this.PlayUpdate.bind(this), -1, this._interval, 0);
		A_WeakLoopSystem.AddUpdater(this._play_loop);
	},
	Stop : function() {
		if (this._play_loop !== undefined) {
			A_WeakLoopSystem.RemoveUpdater(this._play_loop);
			this._play_loop = undefined;
		}
	},
	PlayUpdate : function() {
		let total_count = this._row_count * this._col_count;
		++ this._play_index;
		this._play_index = this._play_index % total_count;
		this.row_index = ALittle.Math_Floor(this._play_index / this._row_count) + 1;
		this.col_index = this._play_index % this._col_count + 1;
	},
}, "ALittle.SpritePlay");

}